### DISCUSSION TITLE
how to get specified property

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1443] [Q&A] [answered]
[demo.txt](https://github.com/eclipse-langium/langium/files/14896696/demo.txt)

IF var1 < 20 THEN
     b := TRUE;
ELSE
     b := FALSE;
END_IF;

this demo.txt is refered from others ,how can i get  var1 property from Expression;
![image](https://github.com/eclipse-langium/langium/assets/26995040/2fb07c0e-b641-494c-9edd-af8c0fcca2fe)

besides ,i am a little confused about this part
![image](https://github.com/eclipse-langium/langium/assets/26995040/ddab299a-cf41-4045-b497-af2f7f6a7e0b)

**Comment by msujew** [comment, marked as answer]
Hey @zrwsmd,

your `Unary_expression` and `Primary_Expression` parser rules should be refined a bit:

```
Power_expression infers Expression:
    Primary_Expression ({infer BinaryExpression.left=current} operator=('**')  right=Primary_Expression)*;

Unary_expression infers Expression:
   operator=('-' | 'NOT') value=Expression;

// primary_expression = constant | enumerated_value | variable | "(" expression ")" | function_name "(" param_assignment {"," param_assignment} ")";
Primary_Expression infers Expression:
Constant
| '(' Expression ')'
| FunctionExpression
| VariableExpression
| Unary_expression;

FunctionExpression infers Expression:
    {infer FunctionExpression} name=Function_name params=Invoke_subrule;

VariableExpression infers Expression:
    {infer VariableExpression} variable=Variable;
```

See [here](https://langium.org/playground?grammar=OYJwhgthYgBAygFwFDIKYDtEgJ4MQFzKwnEkAUAzokmImhJopQNQC8t9jWA%2BgDYBLagEoyJWMgDcqZAGEA9hmpgsRcbADGi5VjaxyABRBoAZgIAeaACY8BWNMDQh%2BA%2BuD6wAPrCOmL1ngAjVx5qEDtgFzcwD28AJTQYqKck%2BWdKAWAMALtEL1gAOUUAB2Mzc1t7R2dBaNjYACF5eT5EjGT3fNkACxgwDTdQ7Aj8gBUBRg6Y4WlkAHoAKkKAV0ZwjU1tRBVmWCs0DT4%2BxAFtWAW55HhM7JsASSqnWGNEZZAlWAxVwKcicgByFj-Tz-AC0-2EAH5YA96NVZgBVJQ3AKwhxPF5vD5fCA-EBENHw1BFDClPwVXLomquFIeTHvSifb6-ZDkBp2GA4Spwp7eADyAySlOq%2BQAEmgKY8QDNUL5yjkpVM6WhXgymbiWeRCU4eIgcMU0DwMJA0LB-gBiCH6a5ZVFS-Ls424blU-IC7Z8F0i7ziyU86WzOX%2BGzBRBDcIYSK1WnPFVYxk4vF-dlhsIRXX6w3Gxhmy3CfRIjK2%2B727yOzle3mwd1C0uwX2VgOobXOPUGo0m2Oq7HM-HIQsomzCnVtrMm7w226NjPt7Om2aThX%2BmdjnP0nsavv-eB3AqjYH-Xf7kEAESPB4AMufEcji9PRx213G1YmWf8ETu9weEeeQQiz1%2Bf5Xl%2BswpuG6YPnOXbxuqSbIP8DQAJqjAAogeADqfJxCeB4nph2GXvhOGzAkSSQZ264Jr2RD-BecQoQAghe-z5P89FMf8JGJJ60buDwaShIO3LQS%2B1GsqRnrkTmFoQtCi4lv6%2Bj-AAdCxLb6Ch5jFIoTBQlCsyLI0zStCosC8TEjIXMgTQtG0SoiR8gTGW0RBzHMsCUN08jLHwViwD8%2BhsnyfIXrmsn6LAACM%2BQAAzCPm3j-KMcQImhrEAGJMfAaHzO56V3AAGgAsihsBeQA7rAiDyLAZIAG5MLVMAZJGsAxbAKh%2BdFYCMlYmSuJQkKsghwUXuaMUsYlDSjeakVWt4ABEyWpQt%2BQLZlF7ZQtBlLD0fQDE8aateZfCMuclx7eAB3pMMrWUR5t3AEQ1yRq0QQ4PQPAaL0V2DEdwD5Ce3mBG9gQfYa337X9j07bA4w5idlmXPDhonQ5jL-UQJ5vHQpwYGMEyGvIJg8FYYB4N4J50IanW6oTgPUyusMo2ZNLuIyoJ7DjJyKCQVnY%2BAPP4-dmPEAC%2B6saMdwlVaMlKeCUIwvYIB1TEsxoirMTow9EZPcgVM4IyPreSARuwEVdjLPQZvwAcihWGbFt8IIlB2xgDuzE7Ltuw72ui%2BQoxe0IPuUDKVwh37j1-KMttaO7of6EHrtxw7%2BmoBbXzW5HuvRxnVtoAn5Cx-boeQrMoom4yItR6yowV28hd59baf6%2BTVfPh8-ujAbhf16bLeGSziOwJzJw5sTVXdKaZN4FZKP8STM-ZxEfxJdLKE8Hy6U8CeDGIZNZqjHyOH5nLBtj2gszn-T1c563XJeW8ZoECxBs8BAlv0M-r-k6EPtX7-R%2BcBb4r2QC2ABXIP6Z1NCAyMBIpQQL-inZecCwH2gBKpJW-p9KwDcnDboQhYCELALABqIAnKu1gPKPy2lcjPB8mgVAVNPqjhQXrAEu9UIH3%2BCfMKsBmGozZmrJhjM0awL1ohRIcAwQsSKooRA3QzTgn4eTWYkiYBsPgf6T28jFHiK0VSCBmi0HaJEZ9WmF9jEcIYqhHgDECgnh4FLGWrETz7lPpaFRn00YAmUdfRgLd5hLCQCAZYAw3jWDhhKPIVlDKaTJJQDIZwrLxOMIkvGahxB8mcBKBJSSMCzGyTwXJaT8lEIwCYJwjJUkF3yZkkgBUBIlNqXjfQABvOwlS4DllwDU9JihlKtBMIgNgGg3jGCwAAX1gPIA0gs0hsABFhK04RgDdBGY0nJWlSl42EAsWYmzinbJabzTpVTYB9LqWIWADF3ZHLya08gHSKlPB6TgS5eNBmmBGWMkAEzEDTNmSkaqIBFn-AKss-Mqz1lsFuTYZp-SMB7NmHC%2B5OzTkvNNhc45iL6mwAUBAYoMAhCKDRSc-GTyzndI5L0nF%2BSvnDNGeMpggK5l0AWQCexOFWIADIVmZBhQSol4RKCkoRfk5FqAhXEtFe0cVrSqXVLpRk65KEACOywyWIvaVSxoNL3nKoGUMn5zKpkzLZSCxZZoAA8AA%2BbhbB%2BVrJGeqzV8rFCSuQK6rVZTFXYoeYoPFDErDwsNRS55XS9VOgNQGjADKTV-JZea4FHL-jWsdaxW1GbErWu4fa-MzwBUjODaG2NnqS0%2BoVZipVsa8WjCcBAHVmKo2cg%2BUa75TLE1mqBfM0FAIgSsQViQaFIz60gAgJ6sdja-VtowHigw8hyo6ndeG3VbzZ3xs7f81lKa%2B3-AWNwuY3CirHxYgWkdbAF1Lq2WW-ZspF3LrDeUrpNb0VzuuUYCYFZZ1NsjeusNm7fnbuTb2sFCwD3nqLZe8I0BnSzs9UiCsK7n3nNnfUnt7K93KMSgUPk7jSExGWGgNgs6DLuVKF%2B50yG9Bxx0HkbwmBVjAoCKrPgRH8iq3CGAEGppFrkFWshhawhVreBMMsDAAw8aPlNAtfjTVwAQB4D1IsGBuB5DaQtAANKtYVkAlPpKyGp6ZQntrIE-bBrkP6Z1hqIAoJQ2wsDIESgCf1b6zTCH%2BE52A6VxOScULOrzAA1Yl3HWgBe8IhqjYbZg%2BYk0LKz1bXPkrxRGp4sW-MYFndMucbB0tC2k-JyAlA2APDqvIAA1oaSgyxAihNaLMYLXGeMJZfUl3F1zUtwEawIULaAssEaa60Ng3Xeuw3gKMB61M1NIyuNsLgTAXDUCIMoebWBWAcDmwwRqVBNvTfYJwLbWBPUHbU38IuaBWgZaGFNxq3g8tSbsGVyr-FnDKZuNd1b9GblXY0DET0K3DtfbuNEfLAO1On0kGe6EKKft-Y%2B4Dog-QhYFBNCV-YWABBmCcDF3z%2BWoL6LAejk4WOQA47iw9jAT2iavYM%2B0MHTBlu7aYIs%2B7pLHsVep-plTrF6KjARHEAoEIydXfZ89gSb2shEAEFYYbIWeP6AlxgWFtO1OhBq3V3jhWIDrdKxztXtWGFhwYirhb1WDetBogQDNK6SPRebJTvXZuNc0Rc%2BQXTjA3DrYMH0RTiu1NKU0yxd3KoqnsG9wprnNxwcLEVu5zisofeR8Mwz1k4fIAo8YLLwbq5TT-Ct%2BFNPEBgtseI6R1AhlbaXaFpNz7M23Kuyr1J%2BnWBYB6Ex-D-33hfuuw70waQleDig6Z6oZAmOSsk2b4gMj7fJ%2Bt9gAtO46UBNPqWqKFCBRVqT8W%2BphaKFNqL%2BX7G%2Bfdd1%2Bb%2BH2GF2ALYAAG1d%2BbRQmfm7vBL8AF15-r8cQf6QdwJ-n5oovliTHOzPqIWW3I-JKNfQXIhEwE7ZnWA5-IQPIAEPfHcdKFiC7DIEwYA1wPGdbH9CA9fdA06THeA5gDbJ-C-RAvZJSFAtKDAtAUg8gz7bfRWf4D-HgAA2YcvJYYHYFVpSfGbXgwWJvP-ZASfFnASWfbwdCAhN6KQ2ABIA0OgXvFvRKFCAqO4dxGLSQ0Q-4dKZZTYLAEAFoEbHjNHJgTHAQJ4PPDNEwNIC8RA3LASS-M0E8PkFiSfBw6gJgwHbfM0dg-QuIePZAdKZwxAogOwkAVJMAtzJKdwqqeQaIn9AEJCFiMGJIsNFuGQgQOQ3Q9CUUO4C8NKcqWQvrbZGI8lVw%2BIzwxw0gvwtghxHgfIwotCLiJQ1MXQ%2BiAwRiCWGo7wuolw98PcQolicTE4PgDI8A9grono4IoJfhJwAQBqPyUcKCKyXKTjHreXPQPqYwAYHgTY3rfISgHAXEFoAQDQA4uXerZAUwi3ZAQ4sw8gE8AQPYsMR41ofIO4mmDALkE4s4wQDQMONyPgeQX7fLMkcoOfBfETefAARVhIWiKlM3rwEAAC9DRISLA58ChCi1oCpESGhET0JESTxESLwUS5hdjB8rjs9oSABSVaUE8EqTLE8wDydEzEsobE4cOADTZSVaXkyZaQF4t42krYz4gnUUmkj4rk8kLBV0DBNSKUAedyf4pyQE8Uo4vQWUgrbwCAHyE4YpVoVXWU6QPBb4pTX40IU4jUi4%2BTV2RkIqQ0gQUEC7QHAbCUguDqd2T03raTW464n4v42084jQIgWUrPL0oIHqe0pUnWMJVUNAIqBgPEKM-0wIWMoEu9QM7PGMjIcMh4oM8wjHEnfQM3SgDQcIYoEZeAGrSs6sygkQMucvOYXUqCNvInSw7HXM6M-HDuDGGuO4LsknMjCsqsgQGsvwvQBaW-Dyesicms2ADTbTecwIBsycq-BaF-UzOs9cxcpswgeCa-FiPcjcpcgEQPBABcxs6g-4F-YI%2BvG8zcufFdfvZ8pc6zWtL1O3GQPBHERYjYaqSrJQZANwKBLWEwcAYAf3eJHSXgNk2AAgXBa-FCNAF-OYaQcCjkDwOC7IVQf1eCsMRC-tA8BWaEcBMChtHCltZ0Xk5CuYAAHUCAACZzRyAYpPBIphAOKuLPByAeB4oFhmLMKqLx0aKGx6LcFmLIoAA2di6-GKUEAATjAAYkCAaA0FkCsBPD6xMHShfx4sUpUrUo0q0p0r0oMv4sEr2REqwuouNA8BrE9CkqYsCAAA4FKlKAB2Qy8gYy3y6yoSuy3KHWEYYCzASgMSiCjwF6YAUGcGL6H6JHHUTGXBAEHixixi5STwa-AAPXvL2QytEuwscv4WBgSs%2Bkhl%2BlSqjnSqE3ICypyvyu3KKqE1ErwUsQitAtKq1kDhyO9hTkoGQvICRH2FNi0GMEVnSmDB4FoRbzmGvwgCKhf2v0oHgAwvsvErKv6ugWGoLHdiqUmrQGmtmvmryEWuWs2uipopjhDhGrGqOrSBOuhBmssBsHOpQvWuut6o8G7jbgesOomuetOvermvkDoUWp0p%2Bocr6r7n2reoCE%2BsWu6FFGutCulwsLLOqpStNhurKqgrABgsakeuBuMAYp4BKtho8EJuJpbxeOAAGgYuMuUpfxYCpu2sgugv90Ro%2Bohpb2QvUnjJbDTl%2BqoW5sagvDBNxlJTZJGsPG4ThOPSF3xq5qJv92uAxLmu5PMHloKm4QaG4XQm4W5USmYhlDFulP2N1LlrNHpJYilpZNlp1oQE5O1vJC2pioVOqGQoZoGkClJsrJBuhD9uYElTwV6A0HKxHhhBHKsKiIAsFmsFkB6lNGWEdNqnkAyBOAajMmaHKzACnjAD8kiPn1XM6nnwIAWlVo8GHKxvjuZqU1BDRIYlBAAC1VrGLyoeAX8LgyMyppd0dkACEQ1MAqpqbYB0J4AGLGLWBRKR70dx7OaPAioLweBZA%2BQioSo9wZ6mKFhr9Z7GKNqFhIRGLhK3JpAF6x6xbV717N7t7RgeBxsZ7Gr97D7j7T7hLhB57B7r6J74A16N6t719Rhd7GLFq8rGKMBGKQBe7RKgA&content=JIMQBAbghgTgjGAPGATABjAFQBIFEByAUAEZgBcAvFgEoCquA3IbgDIDKuJ5VIAgu42b4AIgH1QDIA) for the full grammar

> besides ,i am a little confused about this part

What are you confused about?
